axon-server-connector/src/main/java/org/axonframework/axonserver/connector/query/AxonServerQueryBus.java

```diff
@@ -239,18 +258,48 @@ public void start() {
     public <R> Registration subscribe(@Nonnull String queryName,
                                       @Nonnull Type responseType,
                                       @Nonnull MessageHandler<? super QueryMessage<?, R>> handler) {
+        logger.info("subscribing new query " + queryName);
         Registration localRegistration = localSegment.subscribe(queryName, responseType, handler);
         QueryDefinition queryDefinition = new QueryDefinition(queryName, responseType);
         io.axoniq.axonserver.connector.Registration serverRegistration =
                 axonServerConnectionManager.getConnection(context)
                                            .queryChannel()
                                            .registerQueryHandler(localSegmentAdapter, queryDefinition);
 
-        return new AxonServerRegistration(localRegistration, serverRegistration::cancel);
+        queryHandlerNames.add(queryName);
+        return new AxonServerRegistration(() -> unsubscribe(queryName, localRegistration), serverRegistration::cancel);
+    }
+
+    private boolean unsubscribe(String queryName, Registration localSegmentRegistration) {
+        boolean result = localSegmentRegistration.cancel();
+        if (result) {
+            queryHandlerNames.remove(queryName);
+        }
+        return result;
+    }
+
+    private boolean shouldRunQueryLocally(String queryName) {
+        return localSegmentShortCut && queryHandlerNames.contains(queryName);
+    }
+
+    private <Q, R> CompletableFuture<QueryResponseMessage<R>> tryToRunQueryLocally(
+            @Nonnull QueryMessage<Q, R> queryMessage) {
+        if (shouldRunQueryLocally(queryMessage.getQueryName())) {
+            logger.info("running query locally");
```

Comment:
Suggest logging this on debug level, if needed at all. Also, I would add a bit more information to add context to the log entry.